### PR TITLE
Make styled-components types an optional peer peer dependency

### PR DESCRIPTION
**Pull request by mperrotti**
Closes https://github.com/primer/react/issues/2894

### Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation
- [x] Changes are [SSR compatible](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#ssr-compatibility)
- [x] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

### CODE REVIEW

File: src/CounterLabel/CounterLabel.tsx [hunk 1]
```diff
@@ -34,7 +34,6 @@ const CounterLabel = forwardRef<HTMLSpanElement, CounterLabelProps>(
           )}
           {...props}
           as="span"
-          // @ts-expect-error Box is expecting a divelement, but this component forces a span element
```
Comment (mperrotti) [inline src/CounterLabel/CounterLabel.tsx:37]: This started giving me a type error and I'm not sure why

File: src/Token/TokenBase.tsx [hunk 2]
```diff
@@ -137,6 +137,7 @@ const StyledTokenBase = styled.span<
 
 const TokenBase = React.forwardRef(({onRemove, onKeyDown, id, size = defaultTokenSize, ...rest}, forwardedRef) => {
   return (
+    // @ts-expect-error the possible props in `...rest` would be different depending on the `as` prop
     <StyledTokenBase
```
Comment (mperrotti) [inline src/Token/TokenBase.tsx:141]: This started giving me a type error and I'm not sure why

**Comment by mperrotti** [comment]
This looks like it would be a breaking change if a consumer doesn't have `styled-components` installed. @joshblack - should we batch this into v36 updates? Or did I just do this incorrectly?

**Comment by joshblack** [comment]
@mperrotti good point, I think we'll need to re-target the PR to `next-major` because of that unless there is a way we can get around it 🤔

**Comment by mperrotti** [comment]
Thanks @joshblack ! I updated the base branch.

**Review by joshblack — COMMENTED** [review]
It seems like the consumer test is failing with some random TypeScript errors 🤔 Let me know if you need an other set of 👀 on it!

**Comment by mperrotti** [comment]
@joshblack - these don't seem easy to address, but maybe we could have a quick pairing session?

**Review by joshblack — APPROVED** [review]

**Pull request merged**